onnxruntime/__init__.py

```diff
@@ -10,17 +10,6 @@
 __version__ = "1.6.0"
 __author__ = "Microsoft"
 
-import os
```

Comment:
as i explained in https://github.com/microsoft/onnxruntime/pull/6046#issuecomment-740189908
the os.add_dll_directory() must be done here in __init__.py before importing anything from pybind_state
otherwise it does not work. 
I cloned your fork and confirmed it does not work because you moved this bit of code back to _pybind_state.py

Reply:
steps i did to test your change
1) build python package 
build --config Release --cmake_generator "Visual Studio 16 2019" --use_cuda --build_wheel --skip_submodule_sync --update --build --cuda_version=10.2 --cuda_home="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2" --cudnn_home="C:\cudnn\cuda"
2) install python package produced in build\dist
3) import onnxruntime  and you will encounter below error

C:\>python
Python 3.8.6 (tags/v3.8.6:db45529, Sep 23 2020, 15:52:53) [MSC v.1927 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import onnxruntime as ort
C:\Users\jywu\AppData\Local\Programs\Python\Python38\lib\site-packages\onnxruntime\capi\_pybind_state.py:14: UserWarning: Cannot load onnxruntime.capi. Error: 'DLL load failed while importing onnxruntime_pybind11_state: The specified module could not be found.'.
  warnings.warn("Cannot load onnxruntime.capi. Error: '{0}'.".format(str(e)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\jywu\AppData\Local\Programs\Python\Python38\lib\site-packages\onnxruntime\__init__.py", line 13, in <module>
    from onnxruntime.capi._pybind_state import get_all_providers, get_available_providers, get_device, set_seed, \
ImportError: cannot import name 'get_all_providers' from 'onnxruntime.capi._pybind_state' (C:\Users\jywu\AppData\Local\Programs\Python\Python38\lib\site-packages\onnxruntime\capi\_pybind_state.py)

Reply:
Hi @jywu-msft 
The code referenced in this error message

> C:\Users\jywu\AppData\Local\Programs\Python\Python38\lib\site-packages\onnxruntime\capi_pybind_state.py:14: UserWarning: Cannot load onnxruntime.capi. Error: 'DLL load failed while importing onnxruntime_pybind11_state: The specified module could not be found.'.
> warnings.warn("Cannot load onnxruntime.capi. Error: '{0}'.".format(str(e)))

doesn't exist after my change.

Please review changes in `onnxruntime/python/_pybind_state.py` and make sure you are building the right version of code.

Reply:
as I mentioned, I exactly built your changes from your fork: https://github.com/ivanst0/onnxruntime/tree/windows_python_cuda. 
your code breaks the existing code in master.  you cannot remove the code from `__init__.py`
how did you test your changes? did you built the python package the steps I documented in my previous comment and install it and test it?

Reply:
Yes, I did build/test as you described.
Could you please take a look at "Windows GPU CI Pipeline" which completed successfully on this PR?

Reply:
But our python packaging pipeline failed: https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=143202&view=results

Reply:
I wasn't able to catch this error in an incremental (non-clean) build. Should be fixed now. Please trigger the pipelines again.

Reply:
https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=143274&view=results

Reply:
let me try your latest changes.

Reply:
Thank you guys!

[Build 143274](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=143274&view=results) failed because PR https://github.com/microsoft/onnxruntime/pull/6585 updated this pipeline to CUDA 11, but the PATH environment variable wasn't set correctly, so CUDA 11 DLLs couldn't be found at runtime. I fixed that with [this commit](https://github.com/microsoft/onnxruntime/pull/6436/commits/ad1a589945daada3af8ca9f66f8e74114e9db2c2).

[Build 143390](https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=143390&view=results) completed successfully, but it is not relevant since it was run on `refs/pull/6436/head` (the changes in this PR, but without the latest changes in master) so CUDA 10.2 was used in that build (instead of 11.0).

Could you please start this pipeline again on `refs/pull/6436/merge`?

Reply:
I built your latest branch locally (making sure I had the correct branch). I don't encounter the earlier import error now.
Are your changes only good for Cudnn 8.0 ? I think they don't work for cudnn 7.6.x (failing some build time as well as runtime checks)
Will need to check with the team on whether we need to maintain compatibility with 7.6.x

Reply:
My intention was to support any version of cuDNN (until they change DLL naming scheme) although I tested only with cuDNN 8.
Please note that cuDNN 7.x is not available for CUDA 11.x.
I think you need either to officially drop support for cuDNN 7.x or to have a pipeline which builds/tests with that version.

Reply:
> My intention was to support any version of cuDNN (until they change DLL naming scheme) although I tested only with cuDNN 8.
> Please note that cuDNN 7.x is not available for CUDA 11.x.
> I think you need either to officially drop support for cuDNN 7.x or to have a pipeline which builds/tests with that version.

yes, i was testing with cuDNN 7.6 + CUDA 10.2 combination 
i don't think we can entirely drop CUDA 10.2 (for users building from source), but need to confirm that with the team.

Reply:
I didn't mean dropping CUDA 10.2 (cuDNN 8.0.x works with CUDA 10.2).

Anyways, I made a small fix in `onnxruntime_python.cmake` so 10.2 + 7.6 build should now work fine.

Could you please trigger both regular and packaging pipeline to see if there are any outstanding errors?

Reply:
https://aiinfra.visualstudio.com/Lotus/_build/results?buildId=143528&view=results

Reply:
Python packaging pipeline still fails on master ([this line](https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/azure-pipelines/templates/py-packaging-stage.yml#L789) needs to be updated).

Please queue this pipeline again on `refs/pull/6436/merge` once the build issue is fixed in master.

Reply:
Sorry there are multiple issues going on. Even with your fix, it still won't work. We are resolving the issues.

Reply:
Python packaging pipeline seems to be fixed couple of days ago. Could you please trigger both regular builds and Python packaging pipeline?

Reply:
Thank you @snnn!
Python packaging pipeline completed successfully ([Build 144504](https://dev.azure.com/aiinfra/Lotus/_build/results?buildId=144504&view=results)).